src/main/scala/scalismo/mesh/MeshBoundary.scala

```diff
@@ -0,0 +1,142 @@
+  */
+trait MeshBoundary {
+
+  /**
+    * Check if the point with given index is on the boundary of the underlaying mesh.
+    *
+    * @return True if point is part of the boundary.
+    */
+  def pointIsOnBoundary(id: Int): Boolean
+
+  /**
+    * Check if the line between the two points with given indices is on the boundary of the underlaying mesh.
+    * @note This method does not check weather there exists a line between the two points. You will get false for all
+    *       lines not in the mesh.
+    *
+    * @return True if the line is part of the boundary.
+    */
+  def lineIsOnBoundary(id1: Int, id2: Int): Boolean
+
+}
+
+
+
+/**
+  * The TriangularMeshBoundary can be queried if a triangle has one side in common with the mesh boundary.
+  */
+trait TriangularMeshBoundary extends MeshBoundary {
+
+  /**
+    * Check if the triangle with given index is on the boundary of the underlaying mesh.
+    *
+    * @return True if the triangle has at least one border in common with the boundary.
+    */
+  def triangleIsOnBoundary(id: Int): Boolean
+}
+
+
+
+/**
+  * Implementation of a TriangularMeshBoundary
+  */
+private class BoundaryOfATriangleMesh(private var vertexIsOnBorder: IndexedSeq[Boolean],
+                                      private var lineIsOnBorder: CSCMatrix[Boolean], // NOTE: CSCMatrix is a yet experimental sparse matrix in breeze
+                                      private var triangleIsOnBorder: IndexedSeq[Boolean]) extends TriangularMeshBoundary {
+
+  override def pointIsOnBoundary(id: Int) : Boolean = {
+    return vertexIsOnBorder(id)
+  }
+
+  override def lineIsOnBoundary(id1: Int, id2: Int): Boolean = {
+    return lineIsOnBorder(id1,id2);
+  }
+
+  override def triangleIsOnBoundary(id: Int ) : Boolean = {
+    return triangleIsOnBorder(id)
+  }
+}
+
+
+
+/**
+  * Factory for mesh boundaries.
+  */
+object MeshBoundaryFactory {
+
+  /**
+    * Build boundary index for triangle mesh.
+    *
+    * @param mesh Incoming triangle mesh.
+    * @return Boundary that can be queried for by index for points, lines, and triangles.
+    */
+  def triangularMeshBoundary[D <: Dim](mesh: TriangleMesh[D]): TriangularMeshBoundary = {
+
+    val points = mesh.pointSet.points.toIndexedSeq
+    val nPts = points.length
+
+    val triangles = mesh.triangulation.triangles
+    val nTriangles = triangles.length
+    val linesOfATriangle = Seq((0,1),(1,2),(2,0))
+
+    val pointOnBorder = new Array[Boolean](nPts)
+    val lineOnBorder = new CSCMatrix.Builder[Boolean](nPts,nPts).result()
+    val triangleOnBorder = new Array[Boolean](nTriangles)
+
+    // lines in only a single triangle lie on the border,
+    // lines contained in two triangles are not on a border
+    triangles.map{ triangle =>
+      linesOfATriangle.map{ line =>
+        val v1 = triangle.pointIds(line._1).id
+        val v2 = triangle.pointIds(line._2).id
+        lineOnBorder(v1,v2) = ! lineOnBorder(v1,v2)
```

Comment:
growing a `CSCMatrix` by direct element access is very slow, it constantly reallocates

use:
```
val builder = new CSCMatrix.Builder[Boolean](nPts,nPts)
builder.add(i, j, value)
...
val matrix = builder.result()new CSCMatrix.Builder[Boolean](nPts,nPts)
```

(or change the data structure)